docs/specs/tsd-006-startup-error-behavior.md

```diff
@@ -0,0 +1,105 @@
+values of those options and their semantic meaning. It describes the behavior
+for the different options on handling startup-errors.
+
+Startup-errors are all errors occurring in calls to `Start()`, for inputs and
+service-inputs, or `Connect()`, for outputs. The behaviors described below
+should only be applied in cases where the plugin *explicitly* states that an
+startup-error is *retry-able*. This comprises for example network errors
+indicating that the host or service is not reachable (yet) or external
+resources, like a machine or file, which are not yet available but might become
+available later. To indicate a retry-able startup-error the plugin should return
+a predefined error-type, potentially including an explicit flag. Additionally,
+the plugin might indicate partial connects to signal
+
+## Configuration Options and Behaviors
+
+Telegraf must introduce a unified `startup_error_behavior` configuration option
+for inputs and output plugins which is handled directly by the Telegraf
+infrastructure and is not passed down to the plugins. The setting must be
+available on a per-plugin basis and defines how Telegraf behaves on startup
+errors.
+
+For all listed behaviors Telegraf might retry to startup the plugin for a
+limited number of times during the startup phase before actually processing
+data.
+
+### `error` behavior
+
+If you set the `error` setting for the `startup_error_behavior` option Telegraf
+must fail and exit on startup-errors. This must be the default behavior.
+
+### `retry` behavior
+
+When using the `retry` setting for the `startup_error_behavior` option Telegraf
+must *not* fail on startup-errors and should continue running. Telegraf must
+retry to startup the failed plugin in each gather- or write-cycle, for inputs
+or for outputs respectively, for an unlimited number of times. Neither the
+plugin's `Gather()` nor `Write()` method is called as long as the startup did
+not succeed.
+
+### `retry-partial` behavior
+
+The `retry-partial` setting for the `startup_error_behavior` option is similar
+to the `retry` setting but the plugin's `Gather()` or `Write()` method is
+called as-soon-as the startup was successful or the returned startup-error
+indicates a partially successful connection, e.g. a subset of the given
+endpoints are reachable.
+
+### `ignore` behavior
+
+When using the `ignore` setting for the `startup_error_behavior` option Telegraf
+must *not* fail on startup-errors and should continue running. On startup error,
+Telegraf must ignore the plugin as-if it was not configured at all, i.e. the
+plugin must be completely removed from processing.
+
+### Custom behaviors
+
+Plugins are free to indicate no-startup-error and implement their own startup
+error-handling if necessary. This might or might not make use of the mechanisms
+defined above. However, this should be an exception for cases that cannot be
+handled by the proposed general framework.
```

Comment:
I am not a fan of this section. I know you said this is an exception, but this opens the door to more confusing, unexpected behavior and PRs that say hey the spec says I can basically do what I want.

Reply:
Ok. Will remove once the other discussion items are clarified.